# 1. Исследование надёжности заёмщиков — анализ банковских данных

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#Импорт-библиотеки-Pandas" data-toc-modified-id="Импорт-библиотеки-Pandas-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Импорт библиотеки Pandas</a></span></li><li><span><a href="#Загружаем-данные" data-toc-modified-id="Загружаем-данные-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Загружаем данные</a></span></li><li><span><a href="#Смотрим-иформацию-о-файле" data-toc-modified-id="Смотрим-иформацию-о-файле-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Смотрим иформацию о файле</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Обработка-пропусков" data-toc-modified-id="Обработка-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Обработка пропусков</a></span></li><li><span><a href="#Замена-типа-данных" data-toc-modified-id="Замена-типа-данных-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Замена типа данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Лемматизация" data-toc-modified-id="Лемматизация-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Лемматизация</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Категоризация данных</a></span><ul class="toc-item"><li><span><a href="#Категоризация-данных-по-количеству-детей-в-семье" data-toc-modified-id="Категоризация-данных-по-количеству-детей-в-семье-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Категоризация данных по количеству детей в семье</a></span></li><li><span><a href="#Категоризация-данных-по-семейному-положению" data-toc-modified-id="Категоризация-данных-по-семейному-положению-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Категоризация данных по семейному положению</a></span></li><li><span><a href="#Категоризация-данных-по-уровню-дохода" data-toc-modified-id="Категоризация-данных-по-уровню-дохода-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Категоризация данных по уровню дохода</a></span></li><li><span><a href="#Категоризация-данных-по-цели-кредита" data-toc-modified-id="Категоризация-данных-по-цели-кредита-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Категоризация данных по цели кредита</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Ответы-на-вопросы" data-toc-modified-id="Ответы-на-вопросы-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Ответы на вопросы</a></span><ul class="toc-item"><li><span><a href="#Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Есть ли зависимость между наличием детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок?</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

# Описание проекта

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

# Изучение данных из файла

# Описание данных

* `children` — количество детей в семье
* `days_employed` — общий трудовой стаж в днях
* `dob_years` — возраст клиента в годах
* `education` — уровень образования клиента
* `education_id` — идентификатор уровня образования
* `family_status` — семейное положение
* `family_status_id` — идентификатор семейного положения
* `gender` — пол клиента
* `income_type` — тип занятости
* `debt` — имел ли задолженность по возврату кредитов
* `total_income` — ежемесячный доход
* `purpose` — цель получения кредита

## Загрузка данных

### Импорт библиотеки Pandas

In [1]:
import pandas as pd

### Загружаем данные

In [2]:
data = pd.read_csv('data.csv')

### Смотрим иформацию о файле

In [3]:
data.head(15)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12            NaN                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля  
14  165127.911772         покупка жилой недвижимости

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [5]:
data.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

### Вывод

**обнаружены аномальные значения:**

* *кол-во детей* -1 и 20
* *общий трудовой стаж в днях* - 401755 дней и отрицательные значения
* *возраст клиента в годах* - пропущен возраст (стоит 0)
* *ежемесячный доход* -  максимальный доход 2 млн

## Предобработка данных

### Обработка пропусков

замена значения '-1' на 1 в колонке 'количество детей в семье'

и замена значения '20' на 2 в колонке 'количество детей в семье',

In [6]:
data.loc[data['children'] == 20, 'children'] = 2
data.loc[data['children'] == -1, 'children'] = 1
data['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

избавляемся от отрицательных чисел (берем модуль числа)

In [7]:
data['days_employed'] = data['days_employed'].abs()

пропущенные значения для стажа 

In [8]:
data['days_employed'].isna().sum()

2174

пропущенные значения в доходах

In [9]:
data['total_income'].isna().sum()

2174

нулевые значения в возрасте

In [10]:
(data['dob_years'] == 0).sum()

101

In [11]:
data[data['days_employed'].isna()]['income_type'].value_counts()

сотрудник          1105
компаньон           508
пенсионер           413
госслужащий         147
предприниматель       1
Name: income_type, dtype: int64

высчитываем коэффициент отработанных дней на год, далее возраст клиента умножаем на средний коэффициент по группам

In [12]:
data['ratio_days'] = data[data['days_employed'] !=0 ]['days_employed']/((data['dob_years']-16)*365)
data['days_employed'] = data.groupby('income_type')['ratio_days'].transform(lambda x: x.fillna(x.mean()*data['dob_years']*365))

In [13]:
data['days_employed'].isna().sum()

0

In [14]:
data[data['total_income'].isna()]['income_type'].value_counts()

сотрудник          1105
компаньон           508
пенсионер           413
госслужащий         147
предприниматель       1
Name: income_type, dtype: int64

заполняем пропущенные значения средним по зарплате в группах

In [15]:
data.groupby('income_type')['total_income'].median()

income_type
безработный        131339.751676
в декрете           53829.130729
госслужащий        150447.935283
компаньон          172357.950966
пенсионер          118514.486412
предприниматель    499163.144947
сотрудник          142594.396847
студент             98201.625314
Name: total_income, dtype: float64

In [16]:
data.groupby('income_type')['total_income'].mean()

income_type
безработный        131339.751676
в декрете           53829.130729
госслужащий        170898.309923
компаньон          202417.461462
пенсионер          137127.465690
предприниматель    499163.144947
сотрудник          161380.260488
студент             98201.625314
Name: total_income, dtype: float64

меняю нулевой возраст по группам

In [17]:
data['total_income'] = data.groupby('income_type')['total_income'].transform(lambda x: x.fillna(x.mean()))

In [18]:
data['total_income'].isna().sum()

0

количесво "нулевого" возраста по группам

In [19]:
data[data['dob_years'] == 0]['income_type'].value_counts()

сотрудник      55
компаньон      20
пенсионер      20
госслужащий     6
Name: income_type, dtype: int64

медиана по возрасту для групп клиентов

In [20]:
data.groupby('income_type')['dob_years'].median().astype('int')

income_type
безработный        38
в декрете          39
госслужащий        40
компаньон          39
пенсионер          60
предприниматель    42
сотрудник          39
студент            22
Name: dob_years, dtype: int64

меняем "нулевой" возраст на среднее значение по группам клиентов и проверяю

In [21]:
data['dob_years'] = data.groupby('income_type')['dob_years'].transform(lambda x: x.replace(0, int(x.mean())))

In [22]:
(data['dob_years'] == 0).sum()

0

In [23]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 13 columns):
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
ratio_days          19351 non-null float64
dtypes: float64(3), int64(5), object(5)
memory usage: 2.1+ MB


In [24]:
data.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   21525.000000  21525.000000  21525.000000   
mean       0.479721    9872.831411     43.495238      0.817236   
std        0.755528   68768.801865     12.230322      0.548138   
min        0.000000     -68.663078     19.000000      0.000000   
25%        0.000000       0.134383     34.000000      1.000000   
50%        0.000000       0.336284     43.000000      1.000000   
75%        1.000000      18.658630     53.000000      1.000000   
max        5.000000  620507.891384     75.000000      4.000000   

       family_status_id          debt  total_income    ratio_days  
count      21525.000000  21525.000000  2.152500e+04  19351.000000  
mean           0.972544      0.080883  1.673959e+05      4.379367  
std            1.420324      0.272661  9.790695e+04      9.487276  
min            0.000000      0.000000  2.066726e+04    -68.663078  
25%            0.000000      0.000000  1.077982e+05      0.119822  
50%            0.000000      0.000000  1.519313e+05      0.285193  
75%            1.000000      0.000000  2.024175e+05      0.660965  
max            4.000000      1.000000  2.265604e+06    152.860393

**Вывод**

исправил некорректные значения в колонке 'children' (количество детей в семье)
причина появления 20 детей, наверное в том, что в поле ввода по-умолчанию стоит 0, и при вводе числа 2, его ввели перед нулем не затерев его

заменил 0 в колонке 'dob_years' (возраст клиента в годах) на медиану (42 года)
в колонке 'days_employed' убрал 'минус'
заменил медианами пропущенные значения в колонках 'days_employed' и 'total_income'

### Замена типа данных

для замены типа данных из вещественнго типа в целочисленный использую метод astype() с аргументом 'int'

In [25]:
data['days_employed'] = data['days_employed'].astype('int')
data['total_income'] = data['total_income'].astype('int')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 13 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
ratio_days          19351 non-null float64
dtypes: float64(1), int64(7), object(5)
memory usage: 2.1+ MB


**Вывод**

заменил вещественный тип данных на целочисленный в колонках 'days_employed' и 'total_income' ('общий трудовой стаж в днях' и 'ежемесячный доход')

### Обработка дубликатов

находим количество дубликатов

In [26]:
data.duplicated().sum()

54

приводим столбец 'education' к нижнему регистру

In [27]:
data['education'] = data['education'].str.lower()

data.head(10)

children  days_employed  dob_years education  education_id  \
0         1              0         42    высшее             0   
1         1              0         36   среднее             1   
2         0              0         33   среднее             1   
3         3              0         32   среднее             1   
4         0             25         53   среднее             1   
5         0              0         27    высшее             0   
6         0              0         43    высшее             0   
7         0              0         50   среднее             1   
8         2              0         35    высшее             0   
9         0              0         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose  ratio_days  
0               покупка жилья    0.889112  
1     приобретение автомобиля    0.551343  
2               покупка жилья    0.906273  
3  дополнительное образование    0.706292  
4             сыграть свадьбу   25.195563  
5               покупка жилья    0.230681  
6           операции с жильем    0.292156  
7                 образование    0.012311  
8       на проведение свадьбы    0.999260  
9     покупка жилья для семьи    0.239864

пересчитывае дубликаты

In [28]:
data.duplicated().sum()

71

удаляем дубликаты и перестраиваем индекс

In [29]:
data = data.drop_duplicates().reset_index(drop=True)

проверяем наличие дубликатов

In [30]:
data.duplicated().sum()

0

In [31]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 13 columns):
children            21454 non-null int64
days_employed       21454 non-null int64
dob_years           21454 non-null int64
education           21454 non-null object
education_id        21454 non-null int64
family_status       21454 non-null object
family_status_id    21454 non-null int64
gender              21454 non-null object
income_type         21454 non-null object
debt                21454 non-null int64
total_income        21454 non-null int64
purpose             21454 non-null object
ratio_days          19351 non-null float64
dtypes: float64(1), int64(7), object(5)
memory usage: 2.1+ MB


**Вывод**

Найдено 71 дублирующие записи и удалены с перестроение индекса

### Лемматизация

Импортируем библиотеку pymystem3 

In [32]:
from pymystem3 import Mystem
m = Mystem() 
from collections import Counter

смотрим уникальные значения столбца "цель получения кредита" 

In [33]:
data['purpose'].value_counts()

свадьба                                   791
на проведение свадьбы                     768
сыграть свадьбу                           765
операции с недвижимостью                  675
покупка коммерческой недвижимости         661
операции с жильем                         652
покупка жилья для сдачи                   651
операции с коммерческой недвижимостью     650
покупка жилья                             646
жилье                                     646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          624
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

In [34]:
list_of_lemmas = []
for element in data['purpose']:
    lemma = m.lemmatize(element)
    list_of_lemmas.extend(lemma)

unique_lemmas = Counter(list_of_lemmas)
sorted(unique_lemmas.items(), key = lambda pair: pair[1], reverse=True)

[(' ', 33570),
 ('\n', 21454),
 ('недвижимость', 6351),
 ('покупка', 5897),
 ('жилье', 4460),
 ('автомобиль', 4306),
 ('образование', 4013),
 ('с', 2918),
 ('операция', 2604),
 ('свадьба', 2324),
 ('свой', 2230),
 ('на', 2222),
 ('строительство', 1878),
 ('высокий', 1374),
 ('получение', 1314),
 ('коммерческий', 1311),
 ('для', 1289),
 ('жилой', 1230),
 ('сделка', 941),
 ('дополнительный', 906),
 ('заниматься', 904),
 ('проведение', 768),
 ('сыграть', 765),
 ('сдача', 651),
 ('семья', 638),
 ('собственный', 635),
 ('со', 627),
 ('ремонт', 607),
 ('подержанный', 486),
 ('подержать', 478),
 ('приобретение', 461),
 ('профильный', 436)]

выявлены 4 цели кредита:
     * покупка автомобиля
     * операции с недвижимостью
     * получение образования
     * расходы на свадьбу
    

In [35]:
def lemmas(purpose):
    lemma = m.lemmatize(purpose)
    for element in lemma:
        if element == 'автомобиль':
            return 'автомобиль'
        elif element == 'жилье':
            return 'недвижимость'
        elif element == 'недвижимость':
            return 'недвижимость'
        elif element == 'свадьба':
            return 'свадьба'
        elif element == 'образование':
            return 'образование'

In [36]:
data['purpose_cat'] = data['purpose'].apply(lemmas)

In [37]:
def purpose_id(row):
    if row['purpose_cat'] == 'недвижимость':
        return 0
    if row['purpose_cat'] == 'автомобиль':
        return 1
    if row['purpose_cat'] == 'образование':
        return 2
    if row['purpose_cat'] == 'свадьба':
        return 3

In [38]:
data['purpose_id'] = data.apply(purpose_id, axis=1)

In [39]:
data['purpose_cat'].value_counts()

недвижимость    10811
автомобиль       4306
образование      4013
свадьба          2324
Name: purpose_cat, dtype: int64

### Вывод

Большая часть клиентов банка берет кредит на операции с недвижимостью далее следуют покупка автомобиля и получение образования и в конце списка расходы на свадьбу 

## Категоризация данных

### Категоризация данных по количеству детей в семье

In [40]:
child = data['children'].unique()
child.sort()
column = ['children']
child_comp = pd.DataFrame(data=child , columns=column) 

In [41]:
df = data[data['debt'] == 1]
child_comp ['debt'] = df.groupby('children')['debt'].count()

In [42]:
child_comp['debt'] = child_comp['debt'].fillna(0)
child_comp['debt'] = child_comp['debt'].astype('int')

In [43]:
dff = data[data['debt'] == 0]
child_comp ['no_debt'] = dff.groupby('children')['debt'].count()

In [44]:
child_comp['ratio'] = child_comp['debt'] / (child_comp['no_debt'] + child_comp['debt']) * 100 
child_comp

children  debt  no_debt     ratio
0         0  1063    13028  7.543822
1         1   445     4410  9.165808
2         2   202     1926  9.492481
3         3    27      303  8.181818
4         4     4       37  9.756098
5         5     0        9  0.000000

### Категоризация данных по семейному положению

In [45]:
family = data[['family_status_id','family_status']]
family = family.drop_duplicates().reset_index(drop=True)

In [46]:
famaly_debt = data[data['debt'] == 1]
family ['debt'] = famaly_debt.groupby('family_status_id')['debt'].count()

In [47]:
famaly_no_debt = data[data['debt'] == 0]
family ['no_debt'] = famaly_no_debt.groupby('family_status_id')['debt'].count()

In [48]:
family['ratio'] = family['debt'] / (family['no_debt'] + family['debt'])
family

family_status_id          family_status  debt  no_debt     ratio
0                 0        женат / замужем   931    11408  0.075452
1                 1       гражданский брак   388     3763  0.093471
2                 2         вдовец / вдова    63      896  0.065693
3                 3              в разводе    85     1110  0.071130
4                 4  Не женат / не замужем   274     2536  0.097509

### Категоризация данных по уровню дохода

In [49]:
def income_cat(row):
    if row['total_income'] <= 50000:
        return 'низкий'
    elif 50000 < row['total_income'] <= 100000:
        return 'средний'
    elif 100000 < row['total_income'] < 1000000:
        return 'высокий'
    else:
        return 'сверх высокий'

In [50]:
def income_id(row):
    if row['total_income'] <= 50000:
        return 0
    elif 50000 < row['total_income'] <= 100000:
        return 1
    elif 100000 < row['total_income'] < 1000000:
        return 2
    else:
        return 3

In [51]:
data['income_id'] = data.apply(income_id, axis=1)
data['income_cat'] = data.apply(income_cat, axis=1)

In [52]:
income = data[['income_id','income_cat']].sort_values('income_id')
income = income.drop_duplicates().reset_index(drop=True)

In [53]:
income_debt = data[data['debt'] == 1]
income ['debt'] = income_debt.groupby('income_id')['debt'].count()

In [54]:
income_no_debt = data[data['debt'] == 0]
income ['no_debt'] = income_no_debt.groupby('income_id')['debt'].count()

In [55]:
income ['ratio'] = income ['debt'] / (income ['no_debt'] + income ['debt'])
income

income_id     income_cat  debt  no_debt     ratio
0          0         низкий    23      349  0.061828
1          1        средний   331     3760  0.080909
2          2        высокий  1385    15581  0.081634
3          3  сверх высокий     2       23  0.080000

### Категоризация данных по цели кредита

In [56]:
purpose = data[['purpose_id','purpose_cat']].sort_values('purpose_id')
purpose = purpose.drop_duplicates().reset_index(drop=True)

In [57]:
purpose_debt = data[data['debt'] == 1]
purpose ['debt'] = purpose_debt.groupby('purpose_id')['debt'].count()
purpose_no_debt = data[data['debt'] == 0]
purpose ['no_debt'] = purpose_no_debt.groupby('purpose_id')['debt'].count()
purpose ['ratio'] = purpose ['debt'] / (purpose ['no_debt'] + purpose ['debt'])

purpose

purpose_id   purpose_cat  debt  no_debt     ratio
0           0  недвижимость   782    10029  0.072334
1           1    автомобиль   403     3903  0.093590
2           2   образование   370     3643  0.092200
3           3       свадьба   186     2138  0.080034

### Вывод

Категоризировали данные по следующим признакам:

* по количеству детей
* по семейному положению
* по уровню зарплаты
* по цели кредита

## Ответы на вопросы

### Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [58]:
child_comp.sort_values('ratio')

children  debt  no_debt     ratio
5         5     0        9  0.000000
0         0  1063    13028  7.543822
3         3    27      303  8.181818
1         1   445     4410  9.165808
2         2   202     1926  9.492481
4         4     4       37  9.756098

**Вывод**<br>
Наличие детей ухудшает выплату по кредиту

### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [59]:
family.sort_values('ratio')

family_status_id          family_status  debt  no_debt     ratio
2                 2         вдовец / вдова    63      896  0.065693
3                 3              в разводе    85     1110  0.071130
0                 0        женат / замужем   931    11408  0.075452
1                 1       гражданский брак   388     3763  0.093471
4                 4  Не женат / не замужем   274     2536  0.097509

**Вывод**<br>
лица не состоящие и не состоявшие в официальном браке менее ответственны по выплате кредита

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [60]:
income.sort_values('ratio')

income_id     income_cat  debt  no_debt     ratio
0          0         низкий    23      349  0.061828
3          3  сверх высокий     2       23  0.080000
1          1        средний   331     3760  0.080909
2          2        высокий  1385    15581  0.081634

**Вывод**<br>
Клиенты с уровнем дохода менее 50000 руб., менее склонны к просрочке по выплате кредита

### Как разные цели кредита влияют на его возврат в срок?

In [61]:
purpose.sort_values('ratio')

purpose_id   purpose_cat  debt  no_debt     ratio
0           0  недвижимость   782    10029  0.072334
3           3       свадьба   186     2138  0.080034
2           2   образование   370     3643  0.092200
1           1    автомобиль   403     3903  0.093590

**Вывод**<br>
Клиенты берущие кредит на покупку недвижимости наиболее ответственны по выплате кредита

## Общий вывод

* лучше возвращают кредит клиенты не имеющие детей - 7,5% невозрата кредита в срок, имеющие детей допускают просрочку примерно в 9% случаев.
   
* клиенты состоящие в официальном браке или или состояли в официальном браке задерживают платежи в среднем в 7% случаев, клиенты не состоящие в официальном браке допускают просрочку в 9,5%.
    
* клиенты с уровнем дохода менее 50000 руб. допускают просрочку по кредиту в 6% случаев, у остальных просрочка по кредиту состаляет 8%.

* лучше платят клиенты взявшие кредит на недвижимость - 7% просрочек, хуже те кто взял кредит на автомобиль и образование - чуть более 9% случаев.   
    
**Идеальный клиент для банка не имеет детей, состоит или состоял  в официальном браке, берущий кредит на недвижимость**